ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -110,8 +283,469 @@ void heap_recorder_for_each_live_object(
+  // ensure we have compatible hashes
+  int32_t line_as_int32 = (int32_t) location.line;
+  hash = st_hash(&line_as_int32, sizeof(line_as_int32), hash);
+  return hash;
+}
+
+
+// ==============
+// Heap Stack API
+// ==============
+heap_stack* heap_stack_new(ddog_prof_Slice_Location locations) {
+  uint16_t frames_len = locations.len;
+  if (frames_len > MAX_FRAMES_LIMIT) {
+    // Ensure we never attempt to store a heap stack that goes over the MAX_FRAMES_LIMIT
+    // This should not be happening anyway since MAX_FRAMES_LIMIT should be shared with
+    // the stacktrace construction mechanism
+    frames_len = MAX_FRAMES_LIMIT;
+  }
+  heap_stack *stack = ruby_xcalloc(1, sizeof(heap_stack) + frames_len * sizeof(heap_frame));
+  stack->frames_len = frames_len;
+  for (uint16_t i = 0; i < stack->frames_len; i++) {
+    const ddog_prof_Location *location = &locations.ptr[i];
+    stack->frames[i] = (heap_frame) {
+      .name = ruby_strndup(location->function.name.ptr, location->function.name.len),
+      .filename = ruby_strndup(location->function.filename.ptr, location->function.filename.len),
+      // ddog_prof_Location is a int64_t. We don't expect to have to profile files with more than
+      // 2M lines so this cast should be fairly safe?
+      .line = (int32_t) location->line,
+    };
+  }
+  return stack;
 }
 
-// TODO: Remove when things get implemented
-#pragma GCC diagnostic pop
+void heap_stack_free(heap_stack *stack) {
+  for (uint64_t i = 0; i < stack->frames_len; i++) {
+    heap_frame *frame = &stack->frames[i];
+    ruby_xfree(frame->name);
+    ruby_xfree(frame->filename);
+  }
+  ruby_xfree(stack);
+}
+
+int heap_stack_cmp(heap_stack *st1, heap_stack *st2) {
+  if (st1->frames_len != st2->frames_len) {
+    return (int) (st1->frames_len - st2->frames_len);
+  }
+  for (uint64_t i = 0; i < st1->frames_len; i++) {
+    int cmp = heap_frame_cmp(&st1->frames[i], &st2->frames[i]);
+    if (cmp != 0) {
+      return cmp;
+    }
+  }
+  return 0;
+}
+
+// WARN: Must be kept in-sync with ::ddog_location_slice_hash
+st_index_t heap_stack_hash(heap_stack *stack, st_index_t seed) {
+  st_index_t hash = seed;
+  for (uint64_t i = 0; i < stack->frames_len; i++) {
+    hash = heap_frame_hash(&stack->frames[i], hash);
+  }
+  return hash;
+}
+
+// WARN: Must be kept in-sync with ::heap_stack_hash
+st_index_t ddog_location_slice_hash(ddog_prof_Slice_Location locations, st_index_t seed) {
+  st_index_t hash = seed;
+  for (uint64_t i = 0; i < locations.len; i++) {
+    hash = ddog_location_hash(locations.ptr[i], hash);
+  }
+  return hash;
+}
+
+// ===================
+// Heap Record Key API
+// ===================
+heap_record_key* heap_record_key_new(heap_stack *stack) {
+  heap_record_key *key = ruby_xmalloc(sizeof(heap_record_key));
+  key->type = HEAP_STACK;
+  key->heap_stack = stack;
+  return key;
+}
+
+void heap_record_key_free(heap_record_key *key) {
+  ruby_xfree(key);
+}
```

Comment:
Minor: It looks to me that maybe we could have the `heap_record_key` own the `stack` as well?

E.g. whenever a `heap_record` key gets malloc'd, we know it's because there's a new stack being tracked, and they will also get freed together, so we could make freeing the stack become the responsibility of the key and simplify this a bit I guess.

Reply:
There is a kind of 1:1 mapping between a heap-allocated heap_record_key and a heap_record so whether `stack` is owned by key or record seems a bit equivalent from a "created together, freed together".

IMO, the more flexible nature of `heap_record_key` with its potential non-heap-allocated stacks or even not having a `heap_stack` at all, makes it slightly more confusing to give ownership to. Thus why I preferred to keep ownership to the record and not the key.